src/acp/util/getAgentAccess.ts

```diff
@@ -0,0 +1,147 @@
+ * Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import type { WebId } from "../../interfaces";
+import type { AccessModes } from "../type/AccessModes";
+import type { WithAccessibleAcr } from "../type/WithAccessibleAcr";
+import { getPolicyUrlAll } from "../policy/getPolicyUrlAll";
+import { getThing, getUrlAll, ThingPersisted } from "../..";
+import { internal_getAcr } from "../control.internal";
+import { getAcrPolicyUrlAll } from "../policy/getAcrPolicyUrlAll";
+import { ACP } from "../constants";
+import { getAllowModes } from "../policy/getAllowModes";
+import { getDenyModes } from "../policy/getDenyModes";
+
+/** @hidden */
+function isAgentMatched(
+  acr: WithAccessibleAcr,
+  policy: ThingPersisted,
+  webId: string
+): boolean {
+  // TODO: Proper solution
+  // Finalise, release and use the TypeScript ACP Solid library
+  // internal_getActorAccess in acp_v2:96 doesn't reduce the policies properly
+  // policyAppliesTo in acp_v2:256 assumes that every matcher is an agent matcher
+  //
+  // TODO: Stopgap solution
+  // Implement a simplistic reduce function that
+  // matches policies where the agent is present in the matchers
+  const allOfMatchers = getUrlAll(policy, ACP.allOf)
+    .map((url) => getThing(internal_getAcr(acr), url))
+    .filter((thing): thing is ThingPersisted => thing !== null);
+
+  const allOfMatched = allOfMatchers.every((thing) => {
+    return getUrlAll(thing, ACP.agent).includes(webId);
+  });
+
+  const anyOfMatchers = getUrlAll(policy, ACP.anyOf)
+    .map((url) => getThing(internal_getAcr(acr), url))
+    .filter((thing): thing is ThingPersisted => thing !== null);
+
+  const anyOfMatched = anyOfMatchers.some((thing) => {
+    return getUrlAll(thing, ACP.agent).includes(webId);
+  });
+
+  const noneOfMatchers = getUrlAll(policy, ACP.noneOf)
+    .map((url) => getThing(internal_getAcr(acr), url))
+    .filter((thing): thing is ThingPersisted => thing !== null);
+
+  const noneOfMatched = noneOfMatchers.some((thing) => {
+    return getUrlAll(thing, ACP.agent).includes(webId);
+  });
+
+  return (
+    (allOfMatchers.length === 0 || allOfMatched) &&
+    (anyOfMatchers.length === 0 || anyOfMatched) &&
+    (noneOfMatchers.length === 0 || !noneOfMatched)
+  );
```

Comment:
Should the agent be considered matched if no matchers are attached to the policy ? I know the spec says that a policy must reference at least one matcher to be considered, so is this checked somewhere else ? If so, just adding a comment here saying that it has been checked the policy has at least one matcher would be sufficient.

Reply:
Oh, really good catch! Will fix that right now. It does need an extra condition:

```typescript
allOfMatchers.length + anyOfMatchers.length > 0 &&
```

Reply:
Addressed and tested.